src/v/cluster/rm_stm.cc

```diff
@@ -1814,13 +1806,66 @@ ss::future<raft::stm_snapshot> rm_stm::do_take_local_snapshot(
     }
     auto snapshot_offset = last_applied_offset();
     apply_units.return_all();
+    auto write_lock_units = co_await _state_lock.hold_write_lock();
+    /**
+     * Save aborted snapshots still under the apply lock, we need to wait for
+     * the snapshots to persist as we have to update in memory state after the
+     * snapshot files are safe on disk
+     */
+    co_await ss::max_concurrent_for_each(
+      std::move(aborted_snapshots), 64, [this](abort_snapshot& snapshot) {
+          return save_abort_snapshot(std::move(snapshot));
+      });
 
+    _aborted_tx_state.abort_indexes = std::move(final_abort_indexes);
+    fragmented_vector<tx::tx_range> cleaned_aborted_ranges;
+    cleaned_aborted_ranges.reserve(_aborted_tx_state.aborted.size());
+    // preserve those aborted ranges which were not included in the snapshot
+    std::copy_if(
+      _aborted_tx_state.aborted.begin(),
+      _aborted_tx_state.aborted.end(),
+      std::back_inserter(cleaned_aborted_ranges),
+      [&snapshot_ranges](const tx_range& range) {
+          // only preserve ranges that were not included in the snapshot
+          return !snapshot_ranges.contains(range);
+      });
+    vlog(
+      _ctx_log.debug,
+      "Preserved {} aborted transaction ranges",
+      cleaned_aborted_ranges.size());
+    _aborted_tx_state.aborted = std::move(cleaned_aborted_ranges);
+    stm_snapshot.abort_indexes = _aborted_tx_state.abort_indexes.copy();
+    stm_snapshot.aborted = _aborted_tx_state.aborted.copy();
```

Comment:
This looks suspicious to me - the snapshot should contain the state corresponding to the offset at the time the apply lock was taken, and here it may have already advanced (and therefore aborted may be inconsistent)